ui/dashboards/src/context/TimeRangeStateProvider.tsx

```diff
@@ -26,15 +26,41 @@ export interface TimeRangeProviderProps {
 export function TimeRangeStateProvider(props: TimeRangeProviderProps) {
   const { initialValue, children } = props;
 
-  // Use initialValue to populate state (TODO: Will prob need to expose "setter" API eventually)
-  const [timeRange] = useState(() => {
-    if ('pastDuration' in initialValue) {
-      return toAbsoluteTimeRange(initialValue);
-    }
-    return initialValue;
-  });
+  const defaultTimeRange: AbsoluteTimeRange = toAbsoluteTimeRange(initialValue);
+  const [timeRange, setActiveTimeRange] = useState<AbsoluteTimeRange>(defaultTimeRange);
 
-  const ctx = useMemo(() => ({ timeRange }), [timeRange]);
+  // TODO: add support for passing relative time ranges in setTimeRange, ex: { from: 'now-1h', to: 'now' }
+  const setTimeRange: TimeRangeSetter['setTimeRange'] = useCallback((value: AbsoluteTimeRange) => {
+    setActiveTimeRange(value);
+  }, []);
 
-  return <TimeRangeContext.Provider value={ctx}>{children}</TimeRangeContext.Provider>;
+  const ctx = useMemo(() => ({ timeRange, defaultDuration: initialValue.pastDuration }), [timeRange, initialValue]);
+
+  const setters = useMemo(() => ({ setTimeRange }), [setTimeRange]);
+
+  return (
+    <TimeRangeSetterContext.Provider value={setters}>
+      <TimeRangeContext.Provider value={ctx}>{children}</TimeRangeContext.Provider>
+    </TimeRangeSetterContext.Provider>
```

Comment:
Going to think about it some more, but I wonder if there's some way to do this doesn't involve two different contexts/providers. The plugin runtime stuff definitely makes things more complicated...

(Side note: this is totally cool for now, just thinking out loud.)

Reply:
Yeahh I agree it's confusing, wonder if this could be refactored when we bring in Zustand?